cmd/aws/create_test.go

```diff
@@ -0,0 +1,417 @@
+package aws
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"testing"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/service/ec2"
+	ec2Types "github.com/aws/aws-sdk-go-v2/service/ec2/types"
+	"github.com/aws/aws-sdk-go-v2/service/ssm"
+	ssmTypes "github.com/aws/aws-sdk-go-v2/service/ssm/types"
+	"github.com/stretchr/testify/require"
+)
+
+// Mock definitions
+type mockCredentialsProvider struct {
+	fnRetrieve func(ctx context.Context) (*aws.Credentials, error)
+}
+
+func (m *mockCredentialsProvider) Retrieve(ctx context.Context) (aws.Credentials, error) {
+	if m.fnRetrieve == nil {
+		return aws.Credentials{}, errors.New("not implemented")
+	}
+
+	creds, err := m.fnRetrieve(ctx)
+	if err != nil {
+		return aws.Credentials{}, err
+	}
+	return *creds, nil
+}
+
+type mockInstanceTypesPaginator struct {
+	instanceTypes []ec2Types.InstanceTypeInfo
+	err           error
+	called        bool
+}
+
+func (m *mockInstanceTypesPaginator) HasMorePages() bool {
+	return !m.called
+}
+
+func (m *mockInstanceTypesPaginator) NextPage(ctx context.Context, opts ...func(*ec2.Options)) (*ec2.DescribeInstanceTypesOutput, error) {
+	if m.err != nil {
+		return nil, m.err
+	}
+	m.called = true
+	return &ec2.DescribeInstanceTypesOutput{
+		InstanceTypes: m.instanceTypes,
+	}, nil
+}
+
+type mockSSMClient struct {
+	ssm.Client
+	parameterValue string
+	err            error
+}
+
+func (m *mockSSMClient) GetParameter(ctx context.Context, input *ssm.GetParameterInput, opts ...func(*ssm.Options)) (*ssm.GetParameterOutput, error) {
+	if m.err != nil {
+		return nil, m.err
+	}
+	return &ssm.GetParameterOutput{
+		Parameter: &ssmTypes.Parameter{
+			Value: &m.parameterValue,
+		},
+	}, nil
+}
+
+type mockEC2Client struct {
+	ec2.Client
+	images       []ec2Types.Image
+	architecture string
+	err          error
+}
+
+func (m *mockEC2Client) DescribeImages(ctx context.Context, input *ec2.DescribeImagesInput, opts ...func(*ec2.Options)) (*ec2.DescribeImagesOutput, error) {
+	if m.err != nil {
+		return nil, m.err
+	}
+	if m.architecture == "x86_64" {
+		return &ec2.DescribeImagesOutput{
+			Images: []ec2Types.Image{
+				{
+					Architecture: ec2Types.ArchitectureValues("x86_64"),
+				},
+			},
+		}, nil
+	} else if m.architecture == "" {
+		return &ec2.DescribeImagesOutput{
+			Images: m.images,
+		}, nil
+	} else {
+		return nil, errors.New("unexpected architecture")
+	}
+}
```

Comment:
These are still using the old fashion approach where instead of passing the function, you pass the returned values, making them usable exactly once before you have to reinstantiate them again.

Reply:
I am unclear and would need your help ig 😅

Reply:
What Patrick is saying is that you can define a mock implementation of DescribeImages to the mock struct instead of using this single definition. You can see an example of this on the mockCredentialsProvider. It's a function with the same signature as the Method, but that can be dynamically defined for each test case.

Reply:
Yes, sorry, but exactly what Nathan said. I explained it in detail in one of my previous comments but I think because you edited the file either Github or you marked the issue as "resolved", so the comment is collapsed now.